Always convert truthy booleans to 1

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78957
* __->__ #77122

Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

peterbell10 committed:
Always convert truthy booleans to 1

This changes `c10::convert` to interpret booleans as a raw byte and
convert to the integer 1 or 0 before converting to the final type.
This requires some careful work as the compiler want to optimize with
the assumption that a `bool` will always be 1 or 0.

The added test now passes for most operators which support bools via
conversion, but fails for custom kernels.

[ghstack-poisoned]

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


This changes `c10::convert` to interpret booleans as a raw byte and
convert to the integer 1 or 0 before converting to the final type.
This requires some careful work as the compiler wants to optimize with
the assumption that a `bool` will always be 1 or 0.

The added test now passes for most operators which support bools via
conversion, but fails for custom kernels.

[ghstack-poisoned]

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


This changes `c10::convert` to interpret booleans as a raw byte and
convert to the integer 1 or 0 before converting to the final type.
This requires some careful work as the compiler wants to optimize with
the assumption that a `bool` will always be 1 or 0.

The added test now passes for most operators which support bools via
conversion, but fails for custom kernels.

[ghstack-poisoned]

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


This changes `c10::convert` to interpret booleans as a raw byte and
convert to the integer 1 or 0 before converting to the final type.
This requires some careful work as the compiler wants to optimize with
the assumption that a `bool` will always be 1 or 0.

The added test now passes for most operators which support bools via
conversion, but fails for custom kernels.

[ghstack-poisoned]

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


This changes `c10::convert` to interpret booleans as a raw byte and
convert to the integer 1 or 0 before converting to the final type.
This requires some careful work as the compiler wants to optimize with
the assumption that a `bool` will always be 1 or 0.

The added test now passes for most operators which support bools via
conversion, but fails for custom kernels.

[ghstack-poisoned]

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


This changes `c10::convert` to interpret booleans as a raw byte and
convert to the integer 1 or 0 before converting to the final type.
This requires some careful work as the compiler wants to optimize with
the assumption that a `bool` will always be 1 or 0.

The added test now passes for most operators which support bools via
conversion, but fails for custom kernels.

[ghstack-poisoned]

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


This changes `c10::convert` to interpret booleans as a raw byte and
convert to the integer 1 or 0 before converting to the final type.
This requires some careful work as the compiler wants to optimize with
the assumption that a `bool` will always be 1 or 0.

The added test now passes for most operators which support bools via
conversion, but fails for custom kernels.

[ghstack-poisoned]

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


This changes `c10::convert` to interpret booleans as a raw byte and
convert to the integer 1 or 0 before converting to the final type.
This requires some careful work as the compiler wants to optimize with
the assumption that a `bool` will always be 1 or 0.

The added test now passes for most operators which support bools via
conversion, but fails for custom kernels.

[ghstack-poisoned]

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

mruberry commented on test/test_ops.py:
```diff
@@ -968,6 +970,38 @@ def test_complex_half_reference_testing(self, device, dtype, op):
+            expect = op(sample.input, *sample.args, **sample.kwargs)
+
+            inp = convert_boolean_tensors(sample.input)
+            args = tree_map(convert_boolean_tensors, sample.args)
```
We actually have a transform function the SampleInput class (which could probably be updated to use tree_map) -- but I think it will do what you want here

mruberry commented on test/test_ops.py:
```diff
@@ -968,6 +970,38 @@ def test_complex_half_reference_testing(self, device, dtype, op):
             )
             self.assertEqual(actual, expected, exact_dtype=False)
 
+    @ops(op_db, allowed_dtypes=(torch.bool,))
```
Cool test

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -15960,7 +16042,11 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
-           sample_inputs_func=sample_inputs_msort),
+           sample_inputs_func=sample_inputs_msort,
+           skips=(
+               DecorateInfo(unittest.skip("segfaults on some systems"), 'TestCommon',
```
yikes

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -17799,6 +17893,8 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
                 "test_variant_consistency_jit",
                 dtypes=(torch.float32,),
             ),
+            DecorateInfo(unittest.expectedFailure, 'TestCommon',
```
Nice skips and xfails

mruberry commented:
Hey @peterbell10! This is super cool; I'll have to take a closer look with @ngimel. 

What's up with all the xfails and skips? Are there operations which are not using the approach introduced here yet? Or is this identifying other issues, too?

ngimel commented on test/test_ops.py:
```diff
@@ -968,6 +970,38 @@ def test_complex_half_reference_testing(self, device, dtype, op):
+            x_int = torch.where(x, true_vals, false_vals)
+
+            # Reinterpret the storage as a boolean tensor
+            ret = torch.empty_like(x).set_(
+                x_int.storage()._untyped(),
+                storage_offset=x_int.storage_offset(),
+                size=x_int.shape,
+                stride=x_int.stride())
```
```suggestion
            ret=x_int.view(torch.bool)
```
This won't work if element_size of torch.bool tensor is ever not like element_size of uint8, but neither will storage reinterpret in this case.

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

peterbell10 committed:
Fix lint on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

peterbell10 commented:
> What's up with all the xfails and skips? Are there operations which are not using the approach introduced here yet? Or is this identifying other issues, too?

This fixes `nonzero`and `count_nonzero`; all `TensorIterator` kernels on cpu; and all cuda kernels that use dynamic casting. Which cover over 90% of operators that support bool. The remaining 10% will need follow-up PRs to fix. Some are just custom kernels that need updated to use `c10::load`, others use library code like `unique` uses `cub` and I will need to figure out what to do in those cases.

mruberry commented on aten/src/ATen/native/cuda/jit_utils.cpp:
```diff
@@ -280,10 +280,30 @@ const std::string dynamic_cast_support_literal = R"ESCAPE(
+  };
+
+  template <>
+  struct LoadImpl<bool> {
```
@ngimel and I reviewing together:

Could we see before/after conversion times? Like bool->long, bool->float, and long->float (to see if non-bool conversions are affected)?

mruberry replied:
Could we also see the perf delta on some TensorIterator kernels, like torch.add with bool x long, bool x float, and long x float?

mruberry reviewed (commented):
@ngimel and I both think this looks great @peterbell10! A performance sanity check (maybe those functions would benefit from the inline hint?) and this should be good to go!

peterbell10 committed:
Remove memcpy usage on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

peterbell10 commented:
I've run some simple timeit benchmarks for a performance comparison. I think it's fair to say that for large arrays (1 million elements) the performance is within expected measurement error. For smaller tensors (10,000 elements) I do see a small ~10% slowdown on CPU which isn't present on the int64 conversion. I expect this is because of caching effects making it not memory-bound at these scales. So, the extra ` x != 0` instruction starts to have a measurable performance impact.

| CPU             | Numel | Master (us) | This PR (us) |
|-----------------|-------|-------------|--------------|
| bool.to(int64)  | 1e6   | 230         | 223          |
| bool.to(int64)  | 1e4   | 3.39        | 3.55         |
| bool.to(float)  | 1e6   | 439         | 440          |
| bool.to(float)  | 1e4   | 5.49        | 5.66         |
| int64.to(float) | 1e6   | 288         | 287          |
| int64.to(float) | 1e4   | 4.19        | 4.11         |
| bool \|\| bool  | 1e6   | 19.4        | 19.9         |
| bool \|\| bool  | 1e4   | 1.29        | 1.45         |
| bool + int64    | 1e6   | 669         | 666          |
| bool + int64    | 1e4   | 8.3         | 9.4          |
| bool + float    | 1e6   | 531         | 537          |
| bool + float    | 1e4   | 7.08        | 7.55         |

| CUDA            | Numel | Master (us) | This PR (us) |
|-----------------|-------|-------------|--------------|
| bool.to(int64)  | 1e6   | 37.2        | 38.5         |
| bool.to(int64)  | 1e4   | 4.21        | 4.09         |
| bool.to(float)  | 1e6   | 22.0        | 22.1         |
| bool.to(float)  | 1e4   | 4.2         | 4.09         |
| int64.to(float) | 1e6   | 32.6        | 32.8         |
| int64.to(float) | 1e4   | 4.23        | 4.10         |
| bool \|\| bool  | 1e6   | 4.19        | 4.31         |
| bool \|\| bool  | 1e4   | 4.27        | 4.22         |
| bool + int64    | 1e6   | 105         | 103          |
| bool + int64    | 1e4   | 3.85        | 3.67         |
| bool + float    | 1e6   | 39.2        | 39.5         |
| bool + float    | 1e4   | 3.75        | 3.69         |

peterbell10 committed:
Add new test xfails on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

mruberry reviewed (approved):
Test failures are unrelated -- thanks @peterbell10!

mruberry commented:
@pytorchbot merge

mruberry commented:
@pytorchbot revert -m="broke some jobs, like https://github.com/pytorch/pytorch/runs/6706333043?check_suite_focus=true" -c=nosignal

peterbell10 committed:
More test skips on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -12643,7 +12643,8 @@ def error_inputs_mean(op_info, device, **kwargs):
            supports_fwgrad_bwgrad=True,
            skips=(
-               DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_non_standard_bool_values'),
+               DecorateInfo(unittest.skip("fails on some platforms"),
```
Ideally we would skip this test on just those platforms, since this test seems important for this PR

peterbell10 replied:
@mruberry changed to just disable the xfail on macos-x86.

peterbell10 committed:
Remove xfail for macos-x86 only on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

mruberry commented:
Is the slowtest failure unrelated?

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

peterbell10 commented:
I'm not familiar with NNC so can't really say. Rebased and hopefully it'll go away, if not I should be able to just remove the xfail on the test.

peterbell10 committed:
Update on "Always convert truthy booleans to 1"


Ref #54789

A `bool` has only two valid values, 1 or 0. Any in-memory value
outside of those leads to undefined behavior. So, instead of
`reinterpret_cast`-ing to `bool*` I read any value as `unsigned char`
and convert to a valid `bool`.

[ghstack-poisoned]

peterbell10 commented:
@mruberry the test is now passing after removing the xfail.

mruberry commented:
@pytorchbot merge
